### PR TITLE
issue4242 - added the displayable json/yaml element

### LINKED ISSUE #4242: Make a JsonDisplay component and use it everywhere
There are a bunch of spots in the UI where we display raw data structures: task payloads, hook templates, task definitions, secret values, and so on.

These are currently using the `Code` component.  Some of them use JSON, some use YAML.

Let's replace these with a `JsonDisplay` component that allows users to switch between JSON and YAML with a toggle of some sort, as well as a nice "copy" button to copy the contents (in the current format).

You should be able to find all of the places where this is necessary by grepping for `'\<dump\>'` and `'JSON.stringify'` in `src/ui/`.  Note that, at least for the moment, we do not want to change places where this data is editable (such as the task creator).

### PR DESCRIPTION
**Pull request by cerkiewny**
Github Bug/Issue: Fixes #4242

### CODE REVIEW

**Comment by petemoore** [comment]
@sarah-clements Would you have time to take a look at this? If not no worries, I can try to find another reviewer.

**Comment by sarah-clements** [comment]
> @sarah-clements Would you have time to take a look at this? If not no worries, I can try to find another reviewer.

Sure, I'll take a look at it Thursday or Friday.

File: ui/src/components/JsonDisplay/index.jsx
Comment (sarah-clements) [inline ui/src/components/JsonDisplay/index.jsx:61]: I think it'd be cleaner to rename the variables/prop names so they are more descriptive, rather than naming them generically then adding comments :)

File: ui/src/components/JsonDisplay/index.jsx
Comment (sarah-clements) [inline ui/src/components/JsonDisplay/index.jsx:16]: I think it'd be cleaner and easier to read if you moved this whole component into the `JsonDisplay` component and then move the two helper functions (toYaml and toJson) inline since they only look to be used once.

File: ui/src/components/JsonDisplay/index.jsx
Comment (sarah-clements) [inline ui/src/components/JsonDisplay/index.jsx:43]: I think you can remove the extra brackets after `=>`

File: ui/src/components/JsonDisplay/index.jsx
Comment (sarah-clements) [inline ui/src/components/JsonDisplay/index.jsx:34]: I'd rename this to `YamlJsonSwitch`

Comment (sarah-clements) [reply ui/src/components/JsonDisplay/index.jsx:34]: or `YamlJsonDisplay`

File: ui/src/components/JsonDisplay/index.jsx
Comment (cerkiewny) [reply ui/src/components/JsonDisplay/index.jsx:61]: I reused properties from the Code block so I won't have to change it in all places the JsonDisplay is used

**Comment by cerkiewny** [comment]
@sarah-clements ready for next review round

File: ui/src/components/JsonDisplay/index.jsx
Comment (sarah-clements) [reply ui/src/components/JsonDisplay/index.jsx:54]: That reasoning doesn't make sense to me. You've already had to update files due to the component name changing, so there's no reason not to refactor the prop names. It's cleaner and a better practice to get into.

Comment (sarah-clements) [reply ui/src/components/JsonDisplay/index.jsx:54]: So please either remove the redundant comments or rename props accordingly.

File: ui/src/components/JsonDisplay/index.jsx
Comment (sarah-clements) [inline ui/src/components/JsonDisplay/index.jsx:9]: Fat arrow function has an implicit return, so if you're not doing anything before the return block on line 10 you can remove it and the brackets.

**Review by sarah-clements — CHANGES_REQUESTED** [review]
This is nearly ready! I'll be on PTO until Aug 2nd, so please ask jwhitlock to merge this once you've made the requested changes.

File: ui/src/components/JsonDisplay/index.jsx
Comment (cerkiewny) [reply ui/src/components/JsonDisplay/index.jsx:54]: yes you are right, I thought I just changed the import name will change them in that case

File: ui/src/components/JsonDisplay/index.jsx
Comment (cerkiewny) [reply ui/src/components/JsonDisplay/index.jsx:9]: Yeah I know just usually I add some code to it, than forget to remove return :D

File: ui/src/components/JsonDisplay/index.jsx
Comment (jwhitlock) [reply ui/src/components/JsonDisplay/index.jsx:54]: I agree, if this new component is going to replace all the Code instances, then it is appropriate to rename some properties to better fit the new usage. I'd support removing unused properties as well, if any no longer make sense or are unused by current instances.

**Comment by cerkiewny** [comment]
@jwhitlock can you have a look

**Comment by jwhitlock** [comment]
@cerkiewny I will take a look, but it may be a race between when I have time (next week is better than this one) and when @sarah-clements gets back...

I'm assuming that the code itself is well reviewed, and I'm going to focus on getting it running locally and seeing how the new component behaves.

**Review by sarah-clements — APPROVED** [review]
I think this looks good, but I haven't tested it. A request for future prs - please don't squash your commits when making changes based on feedback. It makes it harder for reviewers to see what has changed from the previous review. @jwhitlock over to you.

**Comment by cerkiewny** [comment]
@sarah-clements sure will try to remember, it is my usual workflow to squash before merge but I guess github does it for me anyhow.

**Comment by sarah-clements** [comment]
> @sarah-clements sure will try to remember, it is my usual workflow to squash before merge but I guess github does it for me anyhow.

You don't need to do that though. Whomever is merging your pr has the option to squash all commits as part of the merge action :)

**Comment by sarah-clements** [comment]
@jwhitlock or @petemoore this is ready to merge.

**Comment by jwhitlock** [comment]
I found these issues by having my local taskcluster UI talk to the GraphQL instance at https://community-tc.services.mozilla.com/, using the directions at [Hacking on the UI](https://github.com/taskcluster/taskcluster/blob/main/dev-docs/development-process.md#hacking-on-the-ui). I think it is useful for testing UI stuff, since so many views require data and we have no React component tests. Let me know if you have trouble setting this up for yourself.

File: ui/src/components/TaskDetailsCard/index.jsx
Comment (cerkiewny) [reply ui/src/components/TaskDetailsCard/index.jsx:210]: will have a look today, probably the property renaming fixup introduced this

File: ui/src/views/Tasks/TaskDefinition/index.jsx
Comment (cerkiewny) [reply ui/src/views/Tasks/TaskDefinition/index.jsx:51]: I have fixed that now, will push fixup

File: ui/src/components/HookForm/index.jsx
Comment (cerkiewny) [reply ui/src/components/HookForm/index.jsx:895]: Editable component is different to display

File: ui/src/components/IndexedEntry/index.jsx
Comment (cerkiewny) [reply ui/src/components/IndexedEntry/index.jsx:193]: I fixed the error, but it seems that the actual definition of object to display is empty object

**Comment by cerkiewny** [comment]
@petemoore Fetching and rebasing fixed most of the CI failures, there is just one from client-rust which shouldn't be related to my changes

**Comment by jwhitlock** [comment]
> @petemoore Fetching and rebasing fixed most of the CI failures, there is just one from client-rust which shouldn't be related to my changes

Agreed, I opened issue #4997 for the client-rust breakage

**Comment by petemoore** [comment]
Many thanks @cerkiewny!

@jwhitlock The generic-worker failures are certainly unrelated to this change, and I suspect the service-index failures are also unrelated.

**Review by jwhitlock — APPROVED** [review]
Sorry for the long pause @cerkiewny, I was focused on a different project for several weeks.

The changes look good, and I'm excited to see this new functionality on the site!

There was one change needed for the Trigger Hook dialog, which I've added as a commit. If  there are others, I'll add some more.

Waiting for the green light from CI...

**Pull request merged**
